engine/access/rpc/backend/backend_stream_transactions.go

```diff
@@ -0,0 +1,184 @@
+		txID:               tx.ID(),
+		txReferenceBlockID: tx.ReferenceBlockID,
+		messageIndex:       counters.NewMonotonousCounter(0),
+		blockWithTx:        nil,
+	}
+
+	sub := subscription.NewHeightBasedSubscription(
+		b.sendBufferSize,
+		nextHeight,
+		b.getTransactionStatusResponse(&txInfo),
+	)
+
+	go subscription.NewStreamer(b.log, b.broadcaster, b.sendTimeout, b.responseLimit, sub).Stream(ctx)
+
+	return sub
+}
+
+// getTransactionStatusResponse creates a function for handling transaction status subscriptions based on new block and
+// previous status state metadata.
+func (b *backendSubscribeTransactions) getTransactionStatusResponse(txInfo *TransactionSubscriptionMetadata) func(context.Context, uint64) (interface{}, error) {
+	return func(ctx context.Context, height uint64) (interface{}, error) {
+		highestHeight, err := b.getHighestHeight(flow.BlockStatusFinalized)
+		if err != nil {
+			return nil, fmt.Errorf("could not get highest height for block %d: %w", height, err)
+		}
+
+		// Fail early if no notification has been received for the given block height.
+		// Note: It's possible for the data to exist in the data store before the notification is
+		// received. This ensures a consistent view is available to all streams.
+		if height > highestHeight {
+			return nil, fmt.Errorf("block %d is not available yet: %w", height, storage.ErrNotFound)
+		}
+
+		if txInfo.blockWithTx == nil {
+			// Check if block contains transaction.
+			err := b.searchForTransactionBlock(height, txInfo)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		// Find the transaction status.
+		var txStatus flow.TransactionStatus
+		if txInfo.blockWithTx == nil {
+			txStatus, err = b.txLocalDataProvider.DeriveUnknownTransactionStatus(txInfo.txReferenceBlockID)
+		} else {
+			blockID := txInfo.blockWithTx.ID()
+
+			if !txInfo.txExecuted {
+				// Check if transaction was executed.
+				err := b.searchForTransactionResult(blockID, txInfo)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			txStatus, err = b.txLocalDataProvider.DeriveTransactionStatus(blockID, txInfo.blockWithTx.Height, txInfo.txExecuted)
+		}
+		if err != nil {
+			if !errors.Is(err, state.ErrUnknownSnapshotReference) {
+				irrecoverable.Throw(ctx, err)
+			}
+			return nil, rpc.ConvertStorageError(err)
+		}
+
+		messageIndex := txInfo.messageIndex.Value()
+		if ok := txInfo.messageIndex.Set(messageIndex + 1); !ok {
+			return nil, status.Errorf(codes.Internal, "the message index has already been incremented to %d", txInfo.messageIndex.Value())
+		}
+
+		return &convert.TransactionSubscribeInfo{
+			ID:           txInfo.txID,
+			Status:       txStatus,
+			MessageIndex: messageIndex,
+		}, nil
+	}
+}
+
+// searchForTransactionBlock searches for the block containing transaction information at the given height.
+// It updates the TransactionSubscriptionMetadata with the found block if the transaction is present.
+func (b *backendSubscribeTransactions) searchForTransactionBlock(
```

Comment:
I think the logic of setting the value on `txInfo` should live in the `getTransactionStatusResponse` to make it easier to track for the reader. Then this method could simply return the header if the tx was in the block, or storage.ErrNotFound if it's not.